File: dotnet/ThirtyFour.UI/WrappedElement.cs
using System.Collections.Generic;
using System.Linq;
using OpenQA.Selenium;
using OpenQA.Selenium.Internal;

namespace ThirtyFour.UI
{
    /// <summary>
    /// Wraps a WebElement to provide specialized functionality
    /// </summary>
    public class WrappedElement
    {
        
        /// <summary>
        /// The wrapped WebElement
        /// </summary>
        public IWebElement Element { get; protected set; }

        /// <summary>
        /// The driver instance used to locate the wrapped element
        /// </summary>
        public IWebDriver Driver { get; protected set; }

        /// <summary>
        /// The parent element or null if no parent exists
        /// </summary>
        public IWebElement ParentElement => GetAncestorElement("*[1]");

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="element">The wrapped web element</param>
        public WrappedElement(IWebElement element)
        {
            Element = element;
            // I know I probably shouldn't do this as it's internal
            Driver = ((IWrapsDriver)element).WrappedDriver;

        }


        /// <summary>
        /// Returns the ancestor of the wrapped element filtered by tagName
        /// </summary>
        /// <param name="tagName">the name of the tage to filter by</param>
        /// <returns>IWebElement the mtaching parent element or null</returns>

        public IWebElement GetAncestorElement(string tagName)
        {
            IWebElement parent = null;
            IReadOnlyCollection<IWebElement> els = Element.FindElements(By.XPath("ancestor::" + tagName));
            if (els.Count > 0)
            {
                parent = els.ElementAt(0);
            }
            return parent;
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/WrappedElement.java
package com.github.danhumphrey.thirtyfour.ui;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.WrapsDriver;


/**
 * The WrappedElement class wraps a <code>WebElement</code> to provide specialized functionality.
 *
 */
public class WrappedElement {

	protected WebElement element;
	protected WebDriver driver;
	
	/**
	 * Constructor
	 * 
	 * @param element the <code>WebElement</code> to wrap
	 */
	public WrappedElement(WebElement element) {
		this.element = element;
		this.driver = ((WrapsDriver)this.element).getWrappedDriver();
	}

	/**
	 * Returns the wrapped element
	 * 
	 * @return the wrapped <code>WebElement</code>
	 */
	public WebElement getElement() {
		return this.element;
	}
	
	/**
	 * Returns the driver
	 * 
	 * @return the <code>WebDriver</code> instance used to locate this element
	 */
	public WebDriver getDriver() {
		return this.driver;
	}
	
	/**
	 * Returns the parent of the wrapped element
	 * 
	 * @return the parent <code>WebElement</code> or <code>null</code> if no parent element exists
	 */
	public WebElement getParentElement() {
		return this.getAncestorElement("*[1]");
	}
	
	/**
	 * Return the ancestor of the wrapped element filtered by <code>tagName</code>
	 * @param tagName the name of the tag to filter by
	 * @return the ancestor <code>WebElement</code> or <code>null</code> if no matching ancestor element exists 
	 */
	public WebElement getAncestorElement(String tagName) {
		WebElement parent = null;
		List<WebElement> els = this.element.findElements(By.xpath("ancestor::" + tagName));
		if(els.size() > 0) {
			parent = els.get(0);
		}
		return parent;
	}
}
